### PR TITLE
Explicit dependencies

### PR DESCRIPTION
**Pull request by dex-sv**

### CODE REVIEW

File: meta_aggregation_api/rest_api/create_app.py [hunk 1]
```diff
@@ -39,14 +36,45 @@ def create_app(config: Config):
...
+    chains = dependencies.ChainsConfig(
+        api_key=config.PUBLIC_KEY,
+        domain=config.PUBLIC_API_DOMAIN,
+    )
+    gas_service = dependencies.GasService(
+        config=config,
+        chains=chains,
+    )
+    providers = dependencies.ProvidersConfig()
+    deps = dependencies.Dependencies(
```
Comment (dex-sv) [inline meta_aggregation_api/rest_api/create_app.py:53]: Dependencies for the lifetime of the FastAPI app

File: meta_aggregation_api/rest_api/routes/gas.py [hunk 2]
```diff
@@ -1,23 +1,20 @@
...
-GAS_CACHE_TTL_SEC = 5
 gas_routes = APIRouter()
 
 
 @gas_routes.get('/{chain_id}', response_model=GasResponse)
 @gas_routes.get('/{chain_id}/', include_in_schema=False)
-@cached(ttl=GAS_CACHE_TTL_SEC, **get_cache_config())
 async def get_prices(
     chain_id: int = Path(..., description='Chain ID'),
+    gas_service: dependencies.GasService = Depends(dependencies.gas_service),
```
Comment (dex-sv) [inline meta_aggregation_api/rest_api/routes/gas.py:14]: Dependency injection

File: meta_aggregation_api/rest_api/dependencies.py [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+from meta_aggregation_api.config.providers import ProvidersConfig
+from meta_aggregation_api.services.chains import ChainsConfig
+from meta_aggregation_api.services.gas_service import GasService
+from meta_aggregation_api.services.limit_orders import LimitOrdersService
+from meta_aggregation_api.services.meta_aggregation_service import (
+    MetaAggregationService,
+)
+
+
+class Dependencies(BaseModel):
```
Comment (dex-sv) [inline meta_aggregation_api/rest_api/dependencies.py:15]: Dependency container

File: meta_aggregation_api/rest_api/dependencies.py [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+
+def aiohttp_session(request: fastapi.Request) -> aiohttp.ClientSession:
+    return _get(request).aiohttp_session
+
+
+def config(request: fastapi.Request) -> Config:
+    return _get(request).config
+
+
+def chains(request: fastapi.Request) -> ChainsConfig:
```
Comment (dex-sv) [inline meta_aggregation_api/rest_api/dependencies.py:50]: callables returning a FastAPI dependency

File: meta_aggregation_api/rest_api/routes/gas.py [hunk 5]
```diff
@@ -1,23 +1,20 @@
...
-from meta_aggregation_api.utils.cache import get_cache_config
+from meta_aggregation_api.rest_api import dependencies
 
-GAS_CACHE_TTL_SEC = 5
 gas_routes = APIRouter()
 
 
 @gas_routes.get('/{chain_id}', response_model=GasResponse)
 @gas_routes.get('/{chain_id}/', include_in_schema=False)
-@cached(ttl=GAS_CACHE_TTL_SEC, **get_cache_config())
```
Comment (dex-sv) [inline meta_aggregation_api/rest_api/routes/gas.py:15]: caching went to the services

File: meta_aggregation_api/services/gas_service.py [hunk 6]
```diff
@@ -18,69 +20,83 @@
...
 
+        self.cached = cached(ttl=5, **get_cache_config(config), noself=True)
 
-@cached(ttl=5, **get_cache_config())
-@retry(retry=retry_if_exception_type(ReadTimeout), stop=3)
-async def get_base_gas_price(chain_id: int) -> int:
-    logger.debug('Getting base gas price for network %s', chain_id)
-    web3_client = Web3Client(get_web3_url(chain_id))
-    return await web3_client.w3.eth.gas_price
+        self.get_gas_prices = self.cached(self.get_gas_prices)
```
Comment (dex-sv) [inline meta_aggregation_api/services/gas_service.py:35]: This way the cache is an instance-level variable, not a global variable.

File: meta_aggregation_api/services/meta_aggregation_service.py [hunk 7]
```diff
@@ -24,419 +29,508 @@
...
-        if allowance < sell_amount:
-            logger.debug('Allowance is not enough, getting approve cost')
-            approve_cost = await get_approve_cost(taker_address, spender_address,
-                                                  erc20_contract)
-            approve_costs_per_provider[provider['name']] = approve_cost
+class MetaAggregationService:
+    def __init__(
+        self,
+        *,
+        config: Config,
```
Comment (dex-sv) [inline meta_aggregation_api/services/meta_aggregation_service.py:36]: All the dependencies of the service are requested by the initializer.

File: meta_aggregation_api/services/meta_aggregation_service.py [hunk 8]
```diff
@@ -24,419 +29,508 @@
...
-    logger.debug('Getting approve cost for owner %s', owner_address)
-    owner_address = Web3.toChecksumAddress(owner_address)
-    spender_address = Web3.toChecksumAddress(spender_address)
-    approve_cost = await erc20_contract.functions.approve(spender_address,
-                                                          2 ** 256 - 1).estimate_gas(
-        {'from': owner_address})
-    return approve_cost
-
-
-async def get_approve_costs_per_provider(
```
Comment (dex-sv) [inline meta_aggregation_api/services/meta_aggregation_service.py:61]: the functions are now the services's methods

File: meta_aggregation_api/providers/openocean_v2/openocean_provider_v2.py [hunk 9]
```diff
@@ -1,27 +1,31 @@
...
+from meta_aggregation_api.utils.errors import (
+    AggregationProviderError,
+    BaseAggregationProviderError,
+)
 from meta_aggregation_api.utils.logger import get_logger
 
 logger = get_logger(__name__)
 
 
 class OpenOceanProviderV2(BaseProvider):
```
Comment (shestakoven) [inline meta_aggregation_api/providers/openocean_v2/openocean_provider_v2.py:26]: why didn't you add `__init__` for this provider?

**Review by safor — APPROVED** [review]
👍 👍 👍

**Comment by shestakoven** [comment]
Did you try to run the app locally? It doesn't work.

File: meta_aggregation_api/providers/kyberswap_v1/kyberswap_provider_v1.py [hunk 10]
```diff
@@ -41,10 +47,26 @@ class KyberSwapProviderV1(BaseProvider):
...
+    def __init__(
+        self,
+        session: aiohttp.ClientSession,
+        config: Config,
+        apm_client: ApmClient,
+        **_,
+    ):
+        super().__init__(session, config, apm_client)
+
+        cached(ttl=30, **get_cache_config(self.config), noself=True)(
```
[the remark this answers is not in the document]
Comment (dex-sv) [reply meta_aggregation_api/providers/kyberswap_v1/kyberswap_provider_v1.py:59]: Good point. Thanks.

**Comment by dex-sv** [comment]
@shestakoven 
> Did you try to run the app locally? It doesn't work.

fixed

File: meta_aggregation_api/providers/kyberswap_v1/kyberswap_provider_v1.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (dex-sv) [reply meta_aggregation_api/providers/kyberswap_v1/kyberswap_provider_v1.py]: Hmm, but we might change the config and remove the key_builder parameter. Being a bit paranoid here probably...

**Comment by shestakoven** [comment]
approved

**Review by shestakoven — APPROVED** [review]

**Comment by dex-sv** [comment]
rebased, squashed

**Pull request merged**